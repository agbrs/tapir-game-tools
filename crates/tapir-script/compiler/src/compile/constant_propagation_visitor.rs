use std::{
    collections::HashMap,
    ops::{BitOr, BitOrAssign},
};

use crate::ast::{Expression, ExpressionKind, Function, Statement, StatementKind, SymbolId};

#[derive(Clone, Copy, Debug)]
enum Constant {
    Int(i32),
    Fix(agb_fixnum::Num<i32, 8>),
    Bool(bool),
}

impl From<&Constant> for ExpressionKind<'_> {
    fn from(value: &Constant) -> Self {
        match *value {
            Constant::Int(i) => ExpressionKind::Integer(i),
            Constant::Fix(num) => ExpressionKind::Fix(num),
            Constant::Bool(b) => ExpressionKind::Bool(b),
        }
    }
}

impl TryFrom<&ExpressionKind<'_>> for Constant {
    type Error = ();

    fn try_from(value: &ExpressionKind<'_>) -> Result<Self, Self::Error> {
        Ok(match value {
            ExpressionKind::Integer(i) => Constant::Int(*i),
            ExpressionKind::Fix(num) => Constant::Fix(*num),
            ExpressionKind::Bool(b) => Constant::Bool(*b),
            _ => return Err(()),
        })
    }
}

#[derive(Clone, Copy, PartialEq, Eq)]
enum ConstantPropagationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for ConstantPropagationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for ConstantPropagationResult {
    type Output = ConstantPropagationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == ConstantPropagationResult::DidSomething
            || rhs == ConstantPropagationResult::DidSomething
        {
            ConstantPropagationResult::DidSomething
        } else {
            ConstantPropagationResult::DidNothing
        }
    }
}

fn constant_propagation(function: &mut Function) -> ConstantPropagationResult {
    let mut constant_symbol = HashMap::new();
    constant_propagation_block(&mut function.statements, &mut constant_symbol)
}

fn constant_propagation_block(
    block: &mut [Statement],
    constant_symbols: &mut HashMap<SymbolId, Constant>,
) -> ConstantPropagationResult {
    block
        .iter_mut()
        .map(|statement| match &mut statement.kind {
            StatementKind::Error
            | StatementKind::Continue
            | StatementKind::Break
            | StatementKind::Nop => ConstantPropagationResult::DidNothing,
            StatementKind::Wait => ConstantPropagationResult::DidNothing,
            StatementKind::Assignment { value, .. } => {
                let did_propagate = constant_propagation_expr(value, constant_symbols);

                let symbol_id = statement.meta.get().unwrap();
                constant_symbols.remove(symbol_id);

                did_propagate
            }
            StatementKind::VariableDeclaration { value, .. } => {
                let did_propagate = constant_propagation_expr(value, constant_symbols);

                if let Ok(constant) = Constant::try_from(&value.kind) {
                    let symbol_id = statement.meta.get().unwrap();
                    constant_symbols.insert(*symbol_id, constant);
                }

                did_propagate
            }
            StatementKind::If {
                condition,
                true_block,
                false_block,
            } => todo!(),
            StatementKind::Loop { block } => todo!(),
            StatementKind::Call { name, arguments } => todo!(),
            StatementKind::Spawn { name, arguments } => todo!(),
            StatementKind::Return { values } => todo!(),
        })
        .reduce(BitOr::bitor)
        .unwrap_or(ConstantPropagationResult::DidNothing)
}

fn constant_propagation_expr(
    expression: &mut Expression,
    constant_symbols: &HashMap<SymbolId, Constant>,
) -> ConstantPropagationResult {
    match &mut expression.kind {
        ExpressionKind::Integer(_)
        | ExpressionKind::Fix(_)
        | ExpressionKind::Bool(_)
        | ExpressionKind::Error
        | ExpressionKind::Nop => ConstantPropagationResult::DidNothing,
        ExpressionKind::BinaryOperation {
            ref mut lhs,
            ref mut rhs,
            ..
        } => {
            constant_propagation_expr(lhs, constant_symbols)
                | constant_propagation_expr(rhs, constant_symbols)
        }
        ExpressionKind::Call { arguments, .. } => arguments
            .iter_mut()
            .map(|expr| constant_propagation_expr(expr, constant_symbols))
            .reduce(BitOr::bitor)
            .unwrap_or(ConstantPropagationResult::DidNothing),
        ExpressionKind::Variable(_) => {
            let symbol: &SymbolId = expression.meta.get().expect("Variable should have symbol");
            if let Some(constant) = constant_symbols.get(symbol) {
                expression.kind = constant.into();
                ConstantPropagationResult::DidSomething
            } else {
                ConstantPropagationResult::DidNothing
            }
        }
    }
}

#[cfg(test)]
mod test {
    use std::fs;

    use insta::{assert_ron_snapshot, glob};

    use crate::{
        compile::{
            loop_visitor::visit_loop_check, symtab_visitor::SymTabVisitor,
            type_visitor::TypeVisitor,
        },
        grammar,
        lexer::Lexer,
        reporting::Diagnostics,
        tokens::FileId,
        CompileSettings,
    };

    use super::*;

    #[test]
    fn constant_propagation_snapshot_tests() {
        glob!("snapshot_tests", "constant_propagation/*.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&input, FileId::new(0));
            let parser = grammar::ScriptParser::new();
            let file_id = FileId::new(0);

            let mut diagnostics = Diagnostics::new(file_id, path.file_name().unwrap(), &input);

            let mut script = parser.parse(file_id, &mut diagnostics, lexer).unwrap();

            let compile_settings = CompileSettings { properties: vec![] };

            let mut symtab_visitor = SymTabVisitor::new(&compile_settings);
            let mut type_visitor =
                TypeVisitor::new(&compile_settings, &script.functions, &mut diagnostics);

            for function in &mut script.functions {
                visit_loop_check(function, &mut diagnostics);
                symtab_visitor.visit_function(function, &mut diagnostics);
                type_visitor.visit_function(
                    function,
                    symtab_visitor.get_symtab(),
                    &mut diagnostics,
                );

                constant_propagation(function);
            }

            assert_ron_snapshot!(script, {
                ".**.span" => "[span]",
                ".**.meta" => "[meta]",
            });
        });
    }
}
