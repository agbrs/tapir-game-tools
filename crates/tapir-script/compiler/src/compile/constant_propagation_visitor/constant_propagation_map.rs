use std::collections::{HashMap, HashSet};

use crate::{
    ast::{ExpressionKind, SymbolId},
    CompileSettings,
};

use super::Constant;

#[derive(Debug, Default)]
pub(crate) struct ConstantPropagationMap {
    map: HashMap<SymbolId, Constant>,
    poisoned: HashSet<SymbolId>,
}

impl ConstantPropagationMap {
    pub fn apply_poisons(&mut self, other: &ConstantPropagationMap) {
        self.poisoned.extend(&other.poisoned);
        self.map
            .retain(|symbol_id, _| !other.poisoned.contains(symbol_id));
    }

    pub fn set(&mut self, symbol_id: SymbolId, value: &ExpressionKind) {
        self.poisoned.insert(symbol_id);

        if let Ok(value) = value.try_into() {
            self.map.insert(symbol_id, value);
        } else {
            self.map.remove(&symbol_id);
        }
    }

    pub fn get(&self, symbol_id: SymbolId) -> Option<Constant> {
        self.map.get(&symbol_id).copied()
    }

    pub fn poison_properties(&mut self, compile_settings: &CompileSettings) {
        self.map.retain(|&symbol_id, _| {
            if compile_settings.is_property(symbol_id) {
                self.poisoned.insert(symbol_id);
                false
            } else {
                true
            }
        });
    }

    pub fn snapshot(&self) -> Self {
        Self {
            map: self.map.clone(),
            poisoned: HashSet::new(),
        }
    }
}
