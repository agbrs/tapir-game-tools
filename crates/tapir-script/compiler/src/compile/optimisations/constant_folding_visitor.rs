use std::{mem, ops::BitOr};

use crate::{
    ast::{BinaryOperator, Expression, ExpressionKind, Function},
    reporting::{CompilerErrorKind, Diagnostics},
};

use super::ConstantOptimisationResult;

pub fn constant_fold(
    function: &mut Function,
    diagnostics: &mut Diagnostics,
) -> ConstantOptimisationResult {
    function
        .statements
        .iter_mut()
        .flat_map(|statement| statement.expressions_mut())
        .map(|expr| fold(expr, diagnostics))
        .reduce(BitOr::bitor)
        .unwrap_or(ConstantOptimisationResult::DidNothing)
}

#[rustfmt::skip]
fn fold(exp: &mut Expression, diagnostics: &mut Diagnostics) -> ConstantOptimisationResult {
    let ExpressionKind::BinaryOperation { lhs, operator, rhs } = &mut exp.kind else {
        return ConstantOptimisationResult::DidNothing;
    };

    let did_something = fold(lhs, diagnostics) | fold(rhs, diagnostics);

    use BinaryOperator as B;
    use ExpressionKind as E;

    exp.kind = match (mem::take(&mut lhs.kind), operator, mem::take(&mut rhs.kind)) {
        // ========================
        // Integer maths operations
        // ========================
        (E::Integer(lhs), B::Add,              E::Integer(rhs)) => E::Integer(lhs + rhs),
        (E::Integer(lhs), B::Sub,              E::Integer(rhs)) => E::Integer(lhs - rhs),
        (E::Integer(lhs), B::Mul,              E::Integer(rhs)) => E::Integer(lhs * rhs),
        (E::Integer(lhs), B::Div | B::RealDiv, E::Integer(rhs)) => E::Integer(lhs / rhs), // FIXME: div_floor
        (E::Integer(lhs), B::Mod | B::RealMod, E::Integer(rhs)) => E::Integer(lhs.rem_euclid(rhs)),

        // ===================
        // Integer comparisons
        // ===================
        (E::Integer(lhs), B::EqEq, E::Integer(rhs)) => E::Bool(lhs == rhs),
        (E::Integer(lhs), B::NeEq, E::Integer(rhs)) => E::Bool(lhs != rhs),
        (E::Integer(lhs), B::Gt,   E::Integer(rhs)) => E::Bool(lhs >  rhs),
        (E::Integer(lhs), B::GtEq, E::Integer(rhs)) => E::Bool(lhs >= rhs),
        (E::Integer(lhs), B::Lt,   E::Integer(rhs)) => E::Bool(lhs <  rhs),
        (E::Integer(lhs), B::LtEq, E::Integer(rhs)) => E::Bool(lhs <= rhs),

        // ===============
        // Fix comparisons
        // ===============
        (E::Fix(lhs), B::EqEq, E::Fix(rhs)) => E::Bool(lhs == rhs),
        (E::Fix(lhs), B::NeEq, E::Fix(rhs)) => E::Bool(lhs != rhs),
        (E::Fix(lhs), B::Gt,   E::Fix(rhs)) => E::Bool(lhs >  rhs),
        (E::Fix(lhs), B::GtEq, E::Fix(rhs)) => E::Bool(lhs >= rhs),
        (E::Fix(lhs), B::Lt,   E::Fix(rhs)) => E::Bool(lhs <  rhs),
        (E::Fix(lhs), B::LtEq, E::Fix(rhs)) => E::Bool(lhs <= rhs),

        // ====================
        // Fix maths operations
        // ====================
        (E::Fix(lhs), B::Add,    E::Fix(rhs)) => E::Fix(lhs + rhs),
        (E::Fix(lhs), B::Sub,    E::Fix(rhs)) => E::Fix(lhs - rhs),
        (E::Fix(lhs), B::FixMul, E::Fix(rhs)) => E::Fix(lhs * rhs),
        (E::Fix(lhs), B::FixDiv, E::Fix(rhs)) => E::Fix(lhs / rhs),

        // ===================
        // add / subtract zero
        // ===================
        (E::Integer(0), B::Add | B::Sub, any) | (any, B::Add | B::Sub, E::Integer(0)) => any,
        (any, B::Add | B::Sub, E::Fix(n)) | (E::Fix(n), B::Add | B::Sub, any) if n == 0.into() => any,

        // ================
        // mulitply by zero
        // ================
        (_, B::Mul, E::Integer(0)) | (E::Integer(0), B::Mul, _) => E::Integer(0),
        (_, B::FixMul, E::Fix(n)) | (E::Fix(n), B::FixMul, _) if n == 0.into() => E::Fix(0.into()),

        // ======================
        // multiply / divide by 1
        // ======================
        (any, B::Mul | B::Div | B::RealDiv, E::Integer(1)) => any,
        (E::Integer(1), B::Mul, any) => any,
        (any, B::FixDiv | B::FixMul, E::Fix(n)) if n == 1.into() => any,
        (E::Fix(n), B::FixDiv | B::FixMul, any) if n == 1.into() => any,

        // ===========
        // Divide by 0
        // ===========
        (_, B::Div | B::RealDiv | B::Mod | B::RealMod, E::Integer(0)) => {
            diagnostics.add_message(CompilerErrorKind::DivideByZero.into_message(rhs.span));
            E::Error
        }
        (_, B::Div | B::RealDiv | B::Mod | B::RealMod, E::Fix(n)) if n == 0.into() => {
            diagnostics.add_message(CompilerErrorKind::DivideByZero.into_message(rhs.span));
            E::Error
        }

        (lhs_kind, _, rhs_kind) => {
            lhs.kind = lhs_kind;
            rhs.kind = rhs_kind;

            return did_something;
        }
    };

    ConstantOptimisationResult::DidSomething
}

#[cfg(test)]
mod test {
    use std::fs;

    use insta::{assert_ron_snapshot, glob};

    use crate::{
        compile::{
            loop_visitor::visit_loop_check, symtab_visitor::SymTabVisitor,
            type_visitor::TypeVisitor,
        },
        grammar,
        lexer::Lexer,
        reporting::Diagnostics,
        tokens::FileId,
        CompileSettings, Property, Type,
    };

    use super::*;

    #[test]
    fn constant_propagation_snapshot_tests() {
        glob!("snapshot_tests", "constant_folding/*.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&input, FileId::new(0));
            let parser = grammar::ScriptParser::new();
            let file_id = FileId::new(0);

            let mut diagnostics = Diagnostics::new(file_id, path.file_name().unwrap(), &input);

            let mut script = parser.parse(file_id, &mut diagnostics, lexer).unwrap();

            let compile_settings = CompileSettings {
                properties: vec![Property {
                    ty: Type::Int,
                    index: 0,
                    name: "int_prop".to_owned(),
                }],
            };

            let mut symtab_visitor = SymTabVisitor::new(&compile_settings);
            let mut type_visitor =
                TypeVisitor::new(&compile_settings, &script.functions, &mut diagnostics);

            for function in &mut script.functions {
                visit_loop_check(function, &mut diagnostics);
                symtab_visitor.visit_function(function, &mut diagnostics);
                type_visitor.visit_function(
                    function,
                    symtab_visitor.get_symtab(),
                    &mut diagnostics,
                );

                constant_fold(function, &mut diagnostics);
            }

            assert_ron_snapshot!(script, {
                ".**.span" => "[span]",
                ".**.meta" => "[meta]",
            });
        });
    }
}
