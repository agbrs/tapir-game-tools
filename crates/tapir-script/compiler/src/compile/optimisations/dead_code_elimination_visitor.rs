use std::collections::HashSet;

use crate::{
    ast::{Expression, ExpressionKind, Function, Statement, StatementKind, SymbolId},
    CompileSettings,
};

#[derive(Default, Clone)]
struct UsedAssignmentsSet {
    used: HashSet<SymbolId>,
    used_declaration: HashSet<SymbolId>,
}

enum DeclarationUsage {
    ValueUsed,
    DeclarationUsed,
    Unused,
}

impl UsedAssignmentsSet {
    fn poison_properties(&mut self, compile_settings: &CompileSettings) {
        self.used.extend(compile_settings.property_symbols());
    }

    fn used(&mut self, symbol: SymbolId) {
        self.used.insert(symbol);
        self.used_declaration.insert(symbol);
    }

    fn remove(&mut self, symbol: SymbolId) -> bool {
        self.used.remove(&symbol)
    }

    fn remove_declaration(&mut self, symbol: SymbolId) -> DeclarationUsage {
        match (
            self.used.remove(&symbol),
            self.used_declaration.remove(&symbol),
        ) {
            (true, true) => DeclarationUsage::ValueUsed,
            (true, false) => unreachable!(),
            (false, true) => DeclarationUsage::DeclarationUsed,
            (false, false) => DeclarationUsage::Unused,
        }
    }

    fn combine(self, other: UsedAssignmentsSet) -> UsedAssignmentsSet {
        UsedAssignmentsSet {
            used: self.used.union(&other.used).copied().collect(),
            used_declaration: self.used_declaration.union(&other.used).copied().collect(),
        }
    }

    fn absorb(&mut self, other: UsedAssignmentsSet) {
        self.used.extend(other.used);
        self.used_declaration.extend(other.used_declaration);
    }
}

pub fn dead_code_eliminate(function: &mut Function, compile_settings: &CompileSettings) {
    eliminate_after_control_flow_diverge(&mut function.statements);

    let mut constant_symbol = UsedAssignmentsSet::default();
    constant_symbol.poison_properties(compile_settings);
    annotate_dead_statements(
        &mut function.statements,
        &mut constant_symbol,
        compile_settings,
        true,
    );

    sweep_dead_statements(&mut function.statements);
}

fn eliminate_after_control_flow_diverge(block: &mut Vec<Statement>) {
    let mut encountered_control_flow_diverge = false;
    block.retain_mut(|statement| {
        if encountered_control_flow_diverge {
            return false;
        }
        match &mut statement.kind {
            StatementKind::Continue | StatementKind::Break => {
                encountered_control_flow_diverge = true;
            }
            StatementKind::Error
            | StatementKind::Nop
            | StatementKind::Wait
            | StatementKind::Assignment { .. }
            | StatementKind::VariableDeclaration { .. }
            | StatementKind::Call { .. }
            | StatementKind::Spawn { .. }
            | StatementKind::Return { .. } => {}
            StatementKind::If {
                true_block,
                false_block,
                ..
            } => {
                eliminate_after_control_flow_diverge(true_block);
                eliminate_after_control_flow_diverge(false_block);
            }
            StatementKind::Loop { block } => {
                eliminate_after_control_flow_diverge(block);
            }
        }
        true
    });
}

fn sweep_dead_statements(block: &mut Vec<Statement>) {
    *block = block
        .drain(..)
        .flat_map(|statement| match statement.kind {
            kind @ (StatementKind::Continue
            | StatementKind::Break
            | StatementKind::Error
            | StatementKind::Nop
            | StatementKind::Wait
            | StatementKind::Call { .. }
            | StatementKind::Spawn { .. }
            | StatementKind::Return { .. }) => Box::new(std::iter::once(Statement {
                span: statement.span,
                kind,
                meta: statement.meta,
            })),
            StatementKind::Assignment { value, ident } => {
                let boxed: Box<dyn Iterator<Item = Statement>> =
                    if statement.meta.get::<DeadStatement>().is_some() {
                        Box::new(extract_side_effects(value).map(move |x| Statement {
                            span: statement.span,
                            kind: x,
                            meta: Default::default(),
                        }))
                    } else {
                        Box::new(std::iter::once(Statement {
                            span: statement.span,
                            kind: StatementKind::Assignment { ident, value },
                            meta: statement.meta,
                        }))
                    };
                boxed
            }
            StatementKind::VariableDeclaration { value, ident } => {
                let boxed: Box<dyn Iterator<Item = Statement>> =
                    if statement.meta.get::<DeadStatement>().is_some() {
                        Box::new(extract_side_effects(value).map(move |x| Statement {
                            span: statement.span,
                            kind: x,
                            meta: Default::default(),
                        }))
                    } else {
                        Box::new(std::iter::once(Statement {
                            span: statement.span,
                            kind: StatementKind::VariableDeclaration { ident, value },
                            meta: statement.meta,
                        }))
                    };
                boxed
            }
            StatementKind::If {
                mut true_block,
                mut false_block,
                condition,
            } => {
                sweep_dead_statements(&mut true_block);
                sweep_dead_statements(&mut false_block);
                Box::new(std::iter::once(Statement {
                    span: statement.span,
                    kind: StatementKind::If {
                        condition,
                        true_block,
                        false_block,
                    },
                    meta: statement.meta,
                }))
            }
            StatementKind::Loop { mut block } => {
                sweep_dead_statements(&mut block);
                Box::new(std::iter::once(Statement {
                    span: statement.span,
                    kind: StatementKind::Loop { block },
                    meta: statement.meta,
                }))
            }
        })
        .collect();
}

#[derive(Debug)]
struct DeadStatement;

fn annotate_dead_statements(
    block: &mut [Statement],
    used_symbols: &mut UsedAssignmentsSet,
    compile_settings: &CompileSettings,
    mark_as_dead: bool,
) {
    for statement in block.iter_mut().rev() {
        match &mut statement.kind {
            StatementKind::Error
            | StatementKind::Continue
            | StatementKind::Break
            | StatementKind::Nop => {}
            StatementKind::Wait => {
                used_symbols.poison_properties(compile_settings);
            }
            StatementKind::Assignment { value, .. } => {
                let symbol = statement.meta.get().unwrap();
                let symbol_is_used = used_symbols.remove(*symbol);
                if !symbol_is_used {
                    if mark_as_dead {
                        statement.meta.set(DeadStatement);
                    }
                } else {
                    dead_code_visit_expression(value, used_symbols, compile_settings);
                }
            }
            StatementKind::VariableDeclaration { value, .. } => {
                let symbol = statement.meta.get().unwrap();
                let symbol_is_used = used_symbols.remove_declaration(*symbol);
                match symbol_is_used {
                    DeclarationUsage::ValueUsed => {
                        dead_code_visit_expression(value, used_symbols, compile_settings);
                    }
                    DeclarationUsage::DeclarationUsed => {
                        if mark_as_dead {
                            value.kind = ExpressionKind::Integer(0);
                        }
                    }
                    DeclarationUsage::Unused => {
                        if mark_as_dead {
                            statement.meta.set(DeadStatement);
                        }
                    }
                }
            }
            StatementKind::If {
                true_block,
                false_block,
                condition,
            } => {
                let mut true_block_used_symbols = used_symbols.clone();
                let mut false_block_used_symbols = used_symbols.clone();
                annotate_dead_statements(
                    true_block,
                    &mut true_block_used_symbols,
                    compile_settings,
                    true,
                );
                annotate_dead_statements(
                    false_block,
                    &mut false_block_used_symbols,
                    compile_settings,
                    true,
                );
                *used_symbols = true_block_used_symbols.combine(false_block_used_symbols);
                dead_code_visit_expression(condition, used_symbols, compile_settings);
            }
            StatementKind::Loop { block } => {
                let mut analyse_existing = used_symbols.clone();
                annotate_dead_statements(block, &mut analyse_existing, compile_settings, false);
                used_symbols.absorb(analyse_existing);
                let mut analyse_existing = used_symbols.clone();
                annotate_dead_statements(block, &mut analyse_existing, compile_settings, true);
                used_symbols.absorb(analyse_existing);
            }
            StatementKind::Call { arguments, .. } => {
                for expr in arguments {
                    dead_code_visit_expression(expr, used_symbols, compile_settings);
                }
                used_symbols.poison_properties(compile_settings);
            }
            StatementKind::Spawn {
                arguments: values, ..
            }
            | StatementKind::Return { values } => {
                for expr in values {
                    dead_code_visit_expression(expr, used_symbols, compile_settings);
                }
            }
        }
    }
}

fn extract_side_effects<'input>(
    expression: Expression<'input>,
) -> Box<dyn Iterator<Item = StatementKind> + '_> {
    match expression.kind {
        ExpressionKind::Integer(_)
        | ExpressionKind::Fix(_)
        | ExpressionKind::Bool(_)
        | ExpressionKind::Variable(_)
        | ExpressionKind::Error
        | ExpressionKind::Nop => Box::new(std::iter::empty()),
        ExpressionKind::BinaryOperation {
            lhs,
            operator: _,
            rhs,
        } => Box::new(extract_side_effects(*lhs).chain(extract_side_effects(*rhs))),
        ExpressionKind::Call { name, arguments } => {
            Box::new(std::iter::once(StatementKind::Call { name, arguments }))
        }
    }
}

fn dead_code_visit_expression(
    expression: &Expression,
    used_symbols: &mut UsedAssignmentsSet,
    compile_settings: &CompileSettings,
) {
    match &expression.kind {
        ExpressionKind::Integer(_)
        | ExpressionKind::Fix(_)
        | ExpressionKind::Bool(_)
        | ExpressionKind::Error
        | ExpressionKind::Nop => {}
        ExpressionKind::BinaryOperation {
            ref lhs, ref rhs, ..
        } => {
            dead_code_visit_expression(lhs, used_symbols, compile_settings);
            dead_code_visit_expression(rhs, used_symbols, compile_settings);
        }
        ExpressionKind::Call { arguments, .. } => {
            for expression in arguments {
                dead_code_visit_expression(expression, used_symbols, compile_settings);
            }
            used_symbols.poison_properties(compile_settings);
        }
        ExpressionKind::Variable(_) => {
            let symbol = expression.meta.get().unwrap();
            used_symbols.used(*symbol);
        }
    }
}

#[cfg(test)]
mod test {
    use std::fs;

    use insta::{assert_ron_snapshot, glob};

    use crate::{
        compile::{
            loop_visitor::visit_loop_check, symtab_visitor::SymTabVisitor,
            type_visitor::TypeVisitor,
        },
        grammar,
        lexer::Lexer,
        reporting::Diagnostics,
        tokens::FileId,
        CompileSettings, Property, Type,
    };

    use super::*;

    #[test]
    fn dead_code_snapshot_tests() {
        glob!("snapshot_tests", "dead_code/*.tapir", |path| {
            let input = fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&input, FileId::new(0));
            let parser = grammar::ScriptParser::new();
            let file_id = FileId::new(0);

            let mut diagnostics = Diagnostics::new(file_id, path.file_name().unwrap(), &input);

            let mut script = parser.parse(file_id, &mut diagnostics, lexer).unwrap();

            let compile_settings = CompileSettings {
                properties: vec![
                    Property {
                        ty: Type::Int,
                        index: 0,
                        name: "int_prop".to_owned(),
                    },
                    Property {
                        ty: Type::Fix,
                        index: 1,
                        name: "fix_prop".to_owned(),
                    },
                    Property {
                        ty: Type::Bool,
                        index: 2,
                        name: "bool_prop".to_owned(),
                    },
                ],
            };

            let mut symtab_visitor = SymTabVisitor::new(&compile_settings);
            let mut type_visitor =
                TypeVisitor::new(&compile_settings, &script.functions, &mut diagnostics);

            for function in &mut script.functions {
                visit_loop_check(function, &mut diagnostics);
                symtab_visitor.visit_function(function, &mut diagnostics);
                type_visitor.visit_function(
                    function,
                    symtab_visitor.get_symtab(),
                    &mut diagnostics,
                );

                dead_code_eliminate(function, &compile_settings);
            }

            assert_ron_snapshot!(script, {
                ".**.span" => "[span]",
                ".**.meta" => "[meta]",
            });
        });
    }
}
