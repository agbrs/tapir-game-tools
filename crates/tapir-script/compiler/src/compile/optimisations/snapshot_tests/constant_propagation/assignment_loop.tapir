var x = 4;
var y = 0;

loop {
    if y == 3 {
        break;
    }

    y = x;
    x = 3;
}

var z = y;