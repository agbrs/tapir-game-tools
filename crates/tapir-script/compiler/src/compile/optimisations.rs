mod constant_folding_visitor;
mod constant_propagation_visitor;
mod dead_code_elimination_visitor;
mod unused_function_visitor;

use std::ops::{BitOr, BitOrAssign};

use constant_folding_visitor::constant_fold;
use constant_propagation_visitor::constant_propagation;
use dead_code_elimination_visitor::dead_code_eliminate;
use unused_function_visitor::unused_function_visitor;

pub(super) use unused_function_visitor::UnusedFunction;

use crate::{
    ast::{ExpressionKind, Function},
    reporting::Diagnostics,
};

use super::CompileSettings;

pub fn optimise(
    functions: &mut [Function],
    compile_settings: &CompileSettings,
    diagnostics: &mut Diagnostics,
) {
    if !compile_settings.enable_optimisations {
        return;
    }

    loop {
        let mut did_something = unused_function_visitor(functions);

        for function in functions.iter_mut() {
            if function.meta.has::<UnusedFunction>() {
                continue; // no point optimising functions which aren't called
            }

            while constant_propagation(function, compile_settings)
                | constant_fold(function, diagnostics)
                | dead_code_eliminate(function, compile_settings)
                == ConstantOptimisationResult::DidSomething
            {
                did_something = ConstantOptimisationResult::DidSomething;
            }
        }

        if did_something == ConstantOptimisationResult::DidNothing {
            break;
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq)]
enum ConstantOptimisationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for ConstantOptimisationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for ConstantOptimisationResult {
    type Output = ConstantOptimisationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == ConstantOptimisationResult::DidSomething
            || rhs == ConstantOptimisationResult::DidSomething
        {
            ConstantOptimisationResult::DidSomething
        } else {
            ConstantOptimisationResult::DidNothing
        }
    }
}

#[derive(Clone, Copy, Debug)]
enum Constant {
    Int(i32),
    Fix(agb_fixnum::Num<i32, 8>),
    Bool(bool),
}

impl From<Constant> for ExpressionKind<'_> {
    fn from(value: Constant) -> Self {
        match value {
            Constant::Int(i) => ExpressionKind::Integer(i),
            Constant::Fix(num) => ExpressionKind::Fix(num),
            Constant::Bool(b) => ExpressionKind::Bool(b),
        }
    }
}

impl TryFrom<&ExpressionKind<'_>> for Constant {
    type Error = ();

    fn try_from(value: &ExpressionKind<'_>) -> Result<Self, Self::Error> {
        Ok(match value {
            ExpressionKind::Integer(i) => Constant::Int(*i),
            ExpressionKind::Fix(num) => Constant::Fix(*num),
            ExpressionKind::Bool(b) => Constant::Bool(*b),
            _ => return Err(()),
        })
    }
}

#[cfg(test)]
mod test {
    use std::fs;

    use insta::{assert_snapshot, glob};

    use crate::{
        compile::{
            loop_visitor::visit_loop_check, symtab_visitor::SymTabVisitor,
            type_visitor::TypeVisitor,
        },
        grammar,
        lexer::Lexer,
        reporting::Diagnostics,
        tokens::FileId,
        CompileSettings,
    };

    use super::*;

    #[test]
    fn optimisation_integration_snapshot_tests() {
        glob!(
            "snapshot_tests",
            "optimisation_integration/*.tapir",
            |path| {
                let input = fs::read_to_string(path).unwrap();

                let lexer = Lexer::new(&input, FileId::new(0));
                let parser = grammar::ScriptParser::new();
                let file_id = FileId::new(0);

                let mut diagnostics = Diagnostics::new(file_id, path.file_name().unwrap(), &input);

                let mut script = parser.parse(file_id, &mut diagnostics, lexer).unwrap();

                let compile_settings = CompileSettings {
                    properties: Vec::new(),
                    enable_optimisations: true,
                };

                let mut symtab_visitor =
                    SymTabVisitor::new(&compile_settings, &mut script.functions, &mut diagnostics);
                let mut type_visitor = TypeVisitor::new(&compile_settings, &script.functions);

                for function in &mut script.functions {
                    visit_loop_check(function, &mut diagnostics);
                    symtab_visitor.visit_function(function, &mut diagnostics);
                    type_visitor.visit_function(
                        function,
                        symtab_visitor.get_symtab(),
                        &mut diagnostics,
                    );
                }

                optimise(&mut script.functions, &compile_settings, &mut diagnostics);

                let pretty_printed = script.pretty_print();

                assert_snapshot!(pretty_printed);
            }
        );
    }
}
