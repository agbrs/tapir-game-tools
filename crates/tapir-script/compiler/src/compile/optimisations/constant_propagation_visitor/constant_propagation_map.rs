use std::collections::{HashMap, HashSet};

use crate::{
    ast::{ExpressionKind, SymbolId},
    CompileSettings,
};

use crate::compile::optimisations::Constant;

#[derive(Debug, Default)]
pub(crate) struct ConstantPropagationMap {
    map: HashMap<SymbolId, Constant>,
    poisoned: HashSet<SymbolId>,
}

impl ConstantPropagationMap {
    pub fn apply_poisons(&mut self, other: &ConstantPropagationMap) {
        self.poisoned.extend(&other.poisoned);
        self.map
            .retain(|symbol_id, _| !other.poisoned.contains(symbol_id));
    }

    pub fn set(&mut self, symbol_id: SymbolId, value: &ExpressionKind) {
        self.poisoned.insert(symbol_id);

        if let Ok(value) = value.try_into() {
            self.map.insert(symbol_id, value);
        } else {
            self.map.remove(&symbol_id);
        }
    }

    pub fn get(&self, symbol_id: SymbolId) -> Option<Constant> {
        self.map.get(&symbol_id).copied()
    }

    pub fn poison_properties(&mut self, compile_settings: &CompileSettings) {
        self.map
            .retain(|&symbol_id, _| !compile_settings.is_property(symbol_id));

        self.poisoned.extend(compile_settings.property_symbols());
    }

    pub fn snapshot(&self) -> Self {
        Self {
            map: self.map.clone(),
            poisoned: HashSet::new(),
        }
    }
}
