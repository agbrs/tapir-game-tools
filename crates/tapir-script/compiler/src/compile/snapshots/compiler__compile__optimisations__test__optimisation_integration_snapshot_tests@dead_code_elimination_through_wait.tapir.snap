---
source: crates/tapir-script/compiler/src/compile/optimisations.rs
expression: pretty_printed
input_file: crates/tapir-script/compiler/src/compile/snapshot_tests/optimisation_integration/dead_code_elimination_through_wait.tapir
---
# @toplevel: {"FunctionId(0)"}

wait;
{
    # {"FunctionId(1)"}
    foo();
}
wait;
wait;

# {"FunctionId(1)"}
fn foo() -> (int,) {
    wait;
    return (33,);
}
